sound/soc/sof/ipc4-pcm.c

```diff
@@ -891,7 +891,7 @@ static snd_pcm_sframes_t sof_ipc4_pcm_delay(struct snd_soc_component *component,
 	tmp_ptr -= time_info->stream_start_offset;
 
 	/* Calculate the delay taking into account that both pointer can wrap */
-	div64_u64_rem(tmp_ptr, substream->runtime->boundary, &tmp_ptr);
+	div64_u64_rem(tmp_ptr, substream->runtime->buffer_size, &tmp_ptr);
```

Comment:
@kv2019i    snd_sof_pcm_platform_pointer()  return the position in [0, buffer_size], we need to wrap tmp_ptr to [0, buffer_size] so that we can compare them. Or after few cycles,  head_ptr is alway < tail_ptr for playback.

Reply:
I dump the log for driver,

[521646.446395] sof-audio-pci-intel-mtl  Link position exceeding buffer size, head 32 tail 16402
[521646.446696] sof-audio-pci-intel-mtl  Link position exceeding buffer size, head 32 tail 16402
[521646.531253] sof-audio-pci-intel-mtl  Link position exceeding buffer size, head 4112 tail 20482
[521646.531503] sof-audio-pci-intel-mtl  Link position exceeding buffer size, head 4112 tail 20482
[521646.616390] sof-audio-pci-intel-mtl  Link position exceeding buffer size, head 8192 tail 24562
[521646.616660] sof-audio-pci-intel-mtl  Link position exceeding buffer size, head 8192 tail 24562
[521646.702375] sof-audio-pci-intel-mtl  Link position exceeding buffer size, head 12320 tail 28690
[521646.702642] sof-audio-pci-intel-mtl  Link position exceeding buffer size, head 12320 tail 28690
[521646.787340] sof-audio-pci-intel-mtl  Link position exceeding buffer size, head 16 tail 32770
[521646.787616] sof-audio-pci-intel-mtl  Link position exceeding buffer size, head 16 tail 32770
[521646.872362] sof-audio-pci-intel-mtl  Link position exceeding buffer size, head 4096 tail 36850
[521646.872613] sof-audio-pci-intel-mtl  Link position exceeding buffer size, head 4096 tail 36850
[521646.958319] sof-audio-pci-intel-mtl  Link position exceeding buffer size, head 8224 tail 40978
[521646.958682] sof-audio-pci-intel-mtl  Link position exceeding buffer size, head 8224 tail 40978
[521647.043383] sof-audio-pci-intel-mtl  Link position exceeding buffer size, head 12304 tail 45058
[521647.043649] sof-audio-pci-intel-mtl  Link position exceeding buffer size, head 12304 tail 45058
[521647.128393] sof-audio-pci-intel-mtl  Link position exceeding buffer size, head 0 tail 49138 
[521647.128650] sof-audio-pci-intel-mtl  Link position exceeding buffer size, head 0 tail 49138 
[521647.214363] sof-audio-pci-intel-mtl  Link position exceeding buffer size, head 4128 tail 53266
[521647.214616] sof-audio-pci-intel-mtl  Link position exceeding buffer size, head 4128 tail 53266
[521647.299364] sof-audio-pci-intel-mtl  Link position exceeding buffer size, head 8208 tail 57346
[521647.299613] sof-audio-pci-intel-mtl  Link position exceeding buffer size, head 8208 tail 57346
[521647.384365] sof-audio-pci-intel-mtl  Link position exceeding buffer size, head 12288 tail 61426
[521647.384584] sof-audio-pci-intel-mtl  Link position exceeding buffer size, head 12288 tail 61426
[521647.470367] sof-audio-pci-intel-mtl  Link position exceeding buffer size, head 32 tail 65554

Reply:
the pointer callback is returning the DMA position between [0 , buffer_size - 1] in frames. The LLP is a linear position in frames, so the LLP needs to be normalized down with buffer_size to be usable.

Reply:
> Right, sorry @RanderWang , I missed you changed also to use the snd_sof_pcm_platform_pointer.
> 
> Hmm, we do lose the ability to express delays longer than buffer_size (and ability to detect such cases reliably, I don't think we in practise need to support such delays with SOF).

Hrm, that is an interesting problem. Indeed we could have a long DSP processing pipeline which would take longer to pass through than the ALSA buffer, it is a valid concern I think
 
> OTOH, so with now, if tmp_ptr is read with get_position, it is a value between 0 and buffersize, while the LLP read case will return a value that will wrap only at UINT64_MAX?